File: templates/python-middleware/src/aport_middleware/middleware.py
"""APort middleware implementation for FastAPI."""

import os
from typing import Optional, Dict, Any, Callable
from fastapi import Request, HTTPException, Depends
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
# from aporthq_sdk import APortClient  # Uncomment when package is available
from .exceptions import APortError, VerificationError


class MockAPortClient:
    """Mock APort Client for template demonstration.
    
    In production, replace with: from aporthq_sdk import APortClient
    """
    
    def __init__(self, api_key: Optional[str] = None, base_url: Optional[str] = None, **kwargs):
        self.api_key = api_key
        self.base_url = base_url or 'https://api.aport.io'
    
    async def verify(self, policy: str, agent_id: str, context: Optional[Dict[str, Any]] = None) -> 'MockVerificationResult':
        """Mock verification - always returns success for template."""
        print(f"[MOCK] Verifying agent {agent_id} against policy {policy}")
        return MockVerificationResult(
            verified=True,
            passport={
                'agentId': agent_id,
                'capabilities': ['read', 'write'],
                'limits': {'requests': 1000, 'period': '1h'}
            },
            policy=policy,
            message="Mock verification successful"
        )


class MockVerificationResult:
    """Mock verification result for template demonstration."""
    
    def __init__(self, verified: bool, passport: Dict[str, Any], policy: str, message: str, details: Optional[Dict[str, Any]] = None):
        self.verified = verified
        self.passport = passport
        self.policy = policy
        self.message = message
        self.details = details or {}


class APortMiddleware:
    """APort middleware for FastAPI applications."""
    
    def __init__(
        self,
        api_key: Optional[str] = None,
        base_url: Optional[str] = None,
        **kwargs
    ):
        """Initialize the APort middleware.
        
        Args:
            api_key: APort API key. If not provided, will use APORT_API_KEY env var.
            base_url: APort API base URL. If not provided, will use APORT_BASE_URL env var.
            **kwargs: Additional client options
        """
        self.client = MockAPortClient(
            api_key=api_key or os.getenv('APORT_API_KEY'),
            base_url=base_url or os.getenv('APORT_BASE_URL', 'https://api.aport.io'),
            **kwargs
        )
    
    def require_policy(
        self,
        policy: str,
        context: Optional[Dict[str, Any]] = None,
        strict: bool = True
    ) -> Callable:
        """Create a dependency that requires a specific policy.
        
        Args:
            policy: Policy pack identifier
            context: Additional context for verification
            strict: Whether to fail on verification errors
            
        Returns:
            FastAPI dependency function
        """
        async def policy_dependency(request: Request) -> Dict[str, Any]:
            """FastAPI dependency that verifies the policy."""
            try:
                # Extract agent ID from request
                agent_id = self._extract_agent_id(request)
                
                if not agent_id:
                    raise HTTPException(
                        status_code=400,
                        detail={
                            'error': 'Agent ID required',
                            'message': 'Agent ID must be provided in headers, query, or body'
                        }
                    )
                
                # Prepare verification context
                verification_context = {
                    'method': request.method,
                    'path': request.url.path,
                    'user_agent': request.headers.get('user-agent'),
                    'ip': request.client.host if request.client else None,
                    **(context or {})
                }
                
                # Verify agent against policy
                result = await self.client.verify(
                    policy=policy,
                    agent_id=agent_id,
                    context=verification_context
                )
                
                if not result.verified:
                    raise HTTPException(
                        status_code=403,
                        detail={
                            'error': 'Verification failed',
                            'message': result.message or 'Agent verification failed',
                            'details': result.details
                        }
                    )
                
                # Return verification result
                return {
                    'verified': True,
                    'passport': result.passport,
                    'policy': policy,
                    'agent_id': agent_id,
                    'result': result
                }
                
            except HTTPException:
                raise
            except Exception as error:
                if strict:
                    raise HTTPException(
                        status_code=500,
                        detail={
                            'error': 'Verification error',
                            'message': 'Internal verification error'
                        }
                    )
                
                # In non-strict mode, return unverified result
                return {
                    'verified': False,
                    'error': str(error)
                }
        
        return policy_dependency
    
    def _extract_agent_id(self, request: Request) -> Optional[str]:
        """Extract agent ID from request.
        
        Args:
            request: FastAPI request object
            
        Returns:
            Agent ID or None if not found
        """
        # Check headers
        agent_id = request.headers.get('x-agent-id') or request.headers.get('x-aport-agent-id')
        if agent_id:
            return agent_id
        
        # Check query parameters
        agent_id = request.query_params.get('agent_id')
        if agent_id:
            return agent_id
        
        # Check JSON body (if available)
        if hasattr(request, '_json') and request._json:
            agent_id = request._json.get('agent_id') or request._json.get('agentId')
            if agent_id:
                return agent_id
        
        return None


# Convenience function for creating policy dependencies
def require_policy(
    policy: str,
    context: Optional[Dict[str, Any]] = None,
    strict: bool = True,
    api_key: Optional[str] = None,
    base_url: Optional[str] = None
) -> Callable:
    """Create a policy dependency without instantiating middleware.
    
    Args:
        policy: Policy pack identifier
        context: Additional context for verification
        strict: Whether to fail on verification errors
        api_key: APort API key
        base_url: APort API base URL
        
    Returns:
        FastAPI dependency function
    """
    middleware = APortMiddleware(api_key=api_key, base_url=base_url)
    return middleware.require_policy(policy, context, strict)

File: templates/nextjs-middleware/middleware.ts
// @ts-ignore
import type { NextRequest } from 'next/server';
// @ts-ignore
import { NextResponse } from 'next/server';

let APortClient: any;
try {
  // Dynamically import APortClient to avoid errors if module is missing in some environments
  // @ts-ignore
  APortClient = require('@aporthq/sdk-node').APortClient;
} catch (e) {
  // Fallback if module is not available (e.g., in local dev without dependency)
  APortClient = undefined;
}
const aportClient = APortClient
  ? new APortClient({
      apiKey: (globalThis as any).process?.env?.APORT_API_KEY,
      baseUrl: (globalThis as any).process?.env?.APORT_BASE_URL || 'https://api.aport.io'
    })
  : undefined;

// Mock client for development/testing
class MockAPortClient {
  async verify(policy: string, agentId: string, context?: any) {
    // Simulate verification logic
    if (agentId === 'test-agent-123') {
      return {
        verified: true,
        passport: {
          agent_id: agentId,
          capabilities: ['payments.refund', 'payments.transfer'],
          limits: {
            refund_amount_max_per_tx: 1000,
            transfer_amount_max_per_tx: 5000
          },
          metadata: {
            verified_at: new Date().toISOString(),
            policy: policy
          }
        },
        message: 'Agent verified successfully'
      };
    }
    
    return {
      verified: false,
      message: 'Agent verification failed',
      details: { error: 'Invalid agent ID' }
    };
  }
}

// Use mock client in development
const client =
  (globalThis as any).process?.env?.NODE_ENV === 'production'
    ? aportClient
    : new MockAPortClient();

export async function middleware(request: NextRequest) {
  // Skip middleware for static files and API routes that don't need verification
  if (
    request.nextUrl.pathname.startsWith('/_next') ||
    request.nextUrl.pathname.startsWith('/favicon.ico') ||
    request.nextUrl.pathname === '/' ||
    request.nextUrl.pathname.startsWith('/api/public')
  ) {
    return NextResponse.next();
  }

  // Extract agent ID from request
  const agentId = 
    request.headers.get('x-agent-id') ||
    request.headers.get('x-aport-agent-id') ||
    request.nextUrl.searchParams.get('agent_id');

  if (!agentId) {
    return NextResponse.json(
      { 
        error: 'Agent ID required',
        message: 'Please provide agent ID in X-Agent-ID header or agent_id query parameter'
      },
      { status: 400 }
    );
  }

  try {
    // Determine policy based on route
    let policy = 'default.access.v1';
    
    if (request.nextUrl.pathname.startsWith('/api/admin')) {
      policy = 'admin.access.v1';
    } else if (request.nextUrl.pathname.startsWith('/api/payments/refund')) {
      policy = 'payments.refund.v1';
    } else if (request.nextUrl.pathname.startsWith('/api/payments/transfer')) {
      policy = 'payments.transfer.v1';
    }

    // Verify agent against policy
    const result = await client.verify(policy, agentId, {
      context: {
        method: request.method,
        path: request.nextUrl.pathname,
        userAgent: request.headers.get('user-agent'),
        ip: request.ip || request.headers.get('x-forwarded-for') || 'unknown'
      }
    });

    if (!result.verified) {
      return NextResponse.json(
        { 
          error: 'Verification failed',
          message: result.message || 'Agent verification failed',
          details: result.details
        },
        { status: 403 }
      );
    }

    // Add verification data to request headers for API routes to access
    const response = NextResponse.next();
    response.headers.set('x-aport-verified', 'true');
    response.headers.set('x-aport-agent-id', agentId);
    response.headers.set('x-aport-policy', policy);
    
    // Store passport data in a header (base64 encoded for complex data)
    if (result.passport) {
      // Use globalThis.Buffer if available (Node.js), otherwise use btoa (browser)
      const passportString = JSON.stringify(result.passport);
      let encodedPassport: string;
      // Use Buffer if available (Node.js), otherwise use btoa (browser)
      if (typeof (globalThis as any).Buffer !== 'undefined') {
        encodedPassport = (globalThis as any).Buffer.from(passportString).toString('base64');
      } else if (typeof btoa !== 'undefined') {
        encodedPassport = btoa(unescape(encodeURIComponent(passportString)));
      } else {
        // Fallback: do not set header if encoding is not possible
        encodedPassport = '';
      }
      if (encodedPassport) {
        response.headers.set('x-aport-passport', encodedPassport);
      }
    }

    return response;

  } catch (error) {
    console.error('APort verification error:', error);
    // In development, allow requests to continue with warning
    const isDev =
      typeof globalThis !== 'undefined' &&
      ((globalThis as any).process?.env?.NODE_ENV === 'development' ||
       (globalThis as any).NODE_ENV === 'development');

    if (isDev) {
      console.warn('APort verification failed, allowing request in development mode');
      return NextResponse.next();
    }

    return NextResponse.json(
      { 
        error: 'Verification error',
        message: 'Internal verification error'
      },
      { status: 500 }
    );
  }
}

// Configure which routes the middleware should run on
export const config = {
  matcher: [
    /*
     * Match all request paths except for the ones starting with:
     * - _next/static (static files)
     * - _next/image (image optimization files)
     * - favicon.ico (favicon file)
     * - public folder
     */
    '/((?!_next/static|_next/image|favicon.ico|.*\\.(?:svg|png|jpg|jpeg|gif|webp)$).*)',
  ],
};
